``Expression is of type "type[int]", not "type[int]"`` when using `assert_type` and `TypeAlias`

<!--
If you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form instead.
Please consider:

- checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
- searching our issue tracker: https://github.com/python/mypy/issues to see if it's already been reported
- asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

<!--
If you're reporting a problem with a specific library function, the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues

If the project you encountered the issue in is open source, please provide a link to the project.
-->

(A clear and concise description of what the bug is.)

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.12&gist=f8dfcc539a523e88dc63f9e6bb423d40

```python
from typing import TypeAlias
from typing_extensions import assert_type

a: TypeAlias = int

assert_type(a, type[int])
```

**Expected Behavior**

<!--
How did you expect mypy to behave? It’s fine if you’re not sure your understanding is correct.
Write down what you thought would happen. If you expected no errors, delete this section.
-->

No issues reported

**Actual Behavior**

```
g.py:6: error: Expression is of type "type[int]", not "type[int]"  [assert-type]
    assert_type(a, type[int])
```

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: mypy g.py
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.12.8

<!-- You can freely edit this text, please remove all the lines you believe are unnecessary. -->

--- a/mypy/subtypes.py
+++ b/mypy/subtypes.py
@@ -1091,7 +1091,12 @@ class SubtypeVisitor(TypeVisitor[bool]):
         right = self.right
         if isinstance(right, TypeType):
             return self._is_subtype(left.item, right.item)
-        if isinstance(right, CallableType):
+        if isinstance(right, (Overloaded, CallableType)):
+            if isinstance(right, Overloaded) and right.is_type_obj():
+                # Same as in other direction: if it's a constructor callable, all
+                # items should belong to the same class' constructor, so it's enough
+                # to check one of them.
+                right = right.items[0]
             if self.proper_subtype and not right.is_type_obj():
                 # We can't accept `Type[X]` as a *proper* subtype of Callable[P, X]
                 # since this will break transitivity of subtyping.

--- a/mypy/subtypes.py
+++ b/mypy/subtypes.py
@@ -1091,12 +1091,12 @@ class SubtypeVisitor(TypeVisitor[bool]):
         right = self.right
         if isinstance(right, TypeType):
             return self._is_subtype(left.item, right.item)
-        if isinstance(right, (Overloaded, CallableType)):
-            if isinstance(right, Overloaded) and right.is_type_obj():
-                # Same as in other direction: if it's a constructor callable, all
-                # items should belong to the same class' constructor, so it's enough
-                # to check one of them.
-                right = right.items[0]
+        if isinstance(right, Overloaded) and right.is_type_obj():
+            # Same as in other direction: if it's a constructor callable, all
+            # items should belong to the same class' constructor, so it's enough
+            # to check one of them.
+            return self._is_subtype(left, right.items[0])
+        if isinstance(right, CallableType):
             if self.proper_subtype and not right.is_type_obj():
                 # We can't accept `Type[X]` as a *proper* subtype of Callable[P, X]
                 # since this will break transitivity of subtyping.

--- a/test-data/unit/check-assert-type-fail.test
+++ b/test-data/unit/check-assert-type-fail.test
@@ -31,3 +31,19 @@ def f(si: arr.array[int]):
 from typing import assert_type, Callable
 def myfunc(arg: int) -> None: pass
 assert_type(myfunc, Callable[[int], None])  # E: Expression is of type "Callable[[Arg(int, 'arg')], None]", not "Callable[[int], None]"
+
+[case testAssertTypeOverload]
+from typing import assert_type, overload
+
+class Foo:
+    @overload
+    def __new__(cls, x: int) -> Foo: ...
+    @overload
+    def __new__(cls, x: str) -> Foo: ...
+    def __new__(cls, x: "int | str") -> Foo:
+        return cls(0)
+
+assert_type(Foo, type[Foo])
+A = Foo
+assert_type(A, type[Foo])
+[builtins fixtures/tuple.pyi]

